Assets/_Game/Scripts/Player/PlayerWeapon.cs

```diff
@@ -48,10 +55,62 @@ public class PlayerWeapon : MonoBehaviour {
   }
 
   public void AddWeapon(IMessage msg) {
-    if (nextAvailableWeaponIndex < weaponPositions.Count && weaponPositions[nextAvailableWeaponIndex] != null) {
-      CreateWeaponAtPosition(weaponPrefab, weaponPositions[nextAvailableWeaponIndex]);
-      nextAvailableWeaponIndex++;
-      CheckButtonWeapon();
+    if (isBuydone == true) {
+      if (nextAvailableWeaponIndex < weaponPositions.Count && weaponPositions[nextAvailableWeaponIndex] != null) {
+        CreateWeaponAtPosition(weaponPrefab, weaponPositions[nextAvailableWeaponIndex]);
+        nextAvailableWeaponIndex++;
+        UpdateWeaponLevelTexts();
+        CheckButtonWeapon();
+        SaveCollectedWeapons();
+      }
+      else {
+        Weapon newWeaponComponent = weaponPrefab.GetComponent<Weapon>();
+        newWeaponComponent.currentWeaponLevel = randomLevel;
+        bool canMerge = false;
+
+        foreach (GameObject weapon in collectedWeapons) {
+          Weapon collectedWeaponComponent = weapon.GetComponent<Weapon>();
+
+          if (collectedWeaponComponent.currentWeaponLevel == newWeaponComponent.currentWeaponLevel) {
+            canMerge = true;
+            collectedWeaponComponent.currentWeaponLevel++;
+            WeaponDataLoader.Ins.LoadWeaponInfo(collectedWeaponComponent.currentWeaponId,
+                collectedWeaponComponent.currentWeaponLevel);
+            if (canMerge) {
+              for (int i = 0; i < weaponInfoButtons.Length; i++) {
+                int position = i;
+                weaponInfoButtons[i].onClick.AddListener(() => UpdateWeaponInfoTexts(position));
```

Comment:
we need to remove the listener before add it again, but I don't think anonymous method will work with RemoveListener